TestBasicReconcilerWithNoExistingEndpoint depends on internet

### What happened?

TestBasicReconcilerWithNoExistingEndpoint will fail if:
1- There is no internet connection
2- The DNS resolution includes ipv6 addresses.

### Steps to reproduce

1. disconnect from the internet
2. make check-unit

### Expected behavior

Unit tests don't depend on external services

### Actual behavior

Unit tests depend on the internet

### Screenshots and logs

No internet:

```
--- FAIL: TestBasicReconcilerWithNoExistingEndpoint (0.01s)
    apiendpointreconciler_test.go:83:
        	Error Trace:	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:83
        	Error:      	Received unexpected error:
        	            	while resolving external address "get.k0s.sh": lookup get.k0s.sh on 192.168.65.7:53: no such host
        	Test:       	TestBasicReconcilerWithNoExistingEndpoint
    apiendpointreconciler_test.go:211:
        	Error Trace:	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:211
        	            				/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:84
        	Error:      	Received unexpected error:
        	            	endpoints "kubernetes" not found
        	Test:       	TestBasicReconcilerWithNoExistingEndpoint
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2e1cd64]

goroutine 108 [running]:
testing.tRunner.func1.2({0x31c2a00, 0x5221e70})
	/usr/local/go/src/testing/testing.go:1526 +0x278
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1529 +0x448
panic({0x31c2a00, 0x5221e70})
	/usr/local/go/src/runtime/panic.go:890 +0x248
github.com/k0sproject/k0s/pkg/component/controller.verifyEndpointAddresses(0x3b8e580?, {0x52379c0, 0x4, 0x4}, {{0x3be0f00, 0xc000897a20}, {0x3b8d000, 0xc000897ad0}, {0x3bc78e8, 0xc0009f4d20}, ...})
	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:212 +0x144
github.com/k0sproject/k0s/pkg/component/controller.TestBasicReconcilerWithNoExistingEndpoint(0x0?)
	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:84 +0x28c
testing.tRunner(0xc000583d40, 0x370f5c0)
	/usr/local/go/src/testing/testing.go:1576 +0x184
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1629 +0x5b8
FAIL	github.com/k0sproject/k0s/pkg/component/controller	0.231s
```

Internet with ipv6 resolution:

```
--- FAIL: TestBasicReconcilerWithNoExistingEndpoint (0.24s)
    apiendpointreconciler_test.go:212:
        	Error Trace:	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:212
        	            				/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:84
        	Error:      	Not equal:
        	            	expected: []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153"}
        	            	actual  : []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153", "64:ff9b::b9c7:6c99", "64:ff9b::b9c7:6d99", "64:ff9b::b9c7:6e99", "64:ff9b::b9c7:6f99"}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-([]string) (len=4) {
        	            	+([]string) (len=8) {
        	            	  (string) (len=15) "185.199.108.153",
        	            	@@ -4,3 +4,7 @@
        	            	  (string) (len=15) "185.199.110.153",
        	            	- (string) (len=15) "185.199.111.153"
        	            	+ (string) (len=15) "185.199.111.153",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6c99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6d99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6e99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6f99"
        	            	 }
        	Test:       	TestBasicReconcilerWithNoExistingEndpoint
--- FAIL: TestBasicReconcilerWithEmptyEndpointSubset (0.00s)
    apiendpointreconciler_test.go:212:
        	Error Trace:	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:212
        	            				/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:119
        	Error:      	Not equal:
        	            	expected: []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153"}
        	            	actual  : []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153", "64:ff9b::b9c7:6c99", "64:ff9b::b9c7:6d99", "64:ff9b::b9c7:6e99", "64:ff9b::b9c7:6f99"}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-([]string) (len=4) {
        	            	+([]string) (len=8) {
        	            	  (string) (len=15) "185.199.108.153",
        	            	@@ -4,3 +4,7 @@
        	            	  (string) (len=15) "185.199.110.153",
        	            	- (string) (len=15) "185.199.111.153"
        	            	+ (string) (len=15) "185.199.111.153",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6c99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6d99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6e99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6f99"
        	            	 }
        	Test:       	TestBasicReconcilerWithEmptyEndpointSubset
--- FAIL: TestReconcilerWithNoNeedForUpdate (0.00s)
    apiendpointreconciler_test.go:212:
        	Error Trace:	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:212
        	            				/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:161
        	Error:      	Not equal:
        	            	expected: []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153"}
        	            	actual  : []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153", "64:ff9b::b9c7:6c99", "64:ff9b::b9c7:6d99", "64:ff9b::b9c7:6e99", "64:ff9b::b9c7:6f99"}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-([]string) (len=4) {
        	            	+([]string) (len=8) {
        	            	  (string) (len=15) "185.199.108.153",
        	            	@@ -4,3 +4,7 @@
        	            	  (string) (len=15) "185.199.110.153",
        	            	- (string) (len=15) "185.199.111.153"
        	            	+ (string) (len=15) "185.199.111.153",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6c99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6d99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6e99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6f99"
        	            	 }
        	Test:       	TestReconcilerWithNoNeedForUpdate
--- FAIL: TestReconcilerWithNeedForUpdate (0.01s)
    apiendpointreconciler_test.go:212:
        	Error Trace:	/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:212
        	            				/go/src/github.com/k0sproject/k0s/pkg/component/controller/apiendpointreconciler_test.go:204
        	Error:      	Not equal:
        	            	expected: []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153"}
        	            	actual  : []string{"185.199.108.153", "185.199.109.153", "185.199.110.153", "185.199.111.153", "64:ff9b::b9c7:6c99", "64:ff9b::b9c7:6d99", "64:ff9b::b9c7:6e99", "64:ff9b::b9c7:6f99"}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-([]string) (len=4) {
        	            	+([]string) (len=8) {
        	            	  (string) (len=15) "185.199.108.153",
        	            	@@ -4,3 +4,7 @@
        	            	  (string) (len=15) "185.199.110.153",
        	            	- (string) (len=15) "185.199.111.153"
        	            	+ (string) (len=15) "185.199.111.153",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6c99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6d99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6e99",
        	            	+ (string) (len=18) "64:ff9b::b9c7:6f99"
        	            	 }
        	Test:       	TestReconcilerWithNeedForUpdate
```

### Additional context

_No response_

--- a/cmd/controller/controller.go
+++ b/cmd/controller/controller.go
@@ -20,6 +20,7 @@ import (
 	"context"
 	"fmt"
 	"io/fs"
+	"net"
 	"os"
 	"os/signal"
 	"path/filepath"
@@ -379,6 +380,7 @@ func (c *command) start(ctx context.Context) error {
 			c.NodeConfig,
 			leaderElector,
 			adminClientFactory,
+			net.DefaultResolver,
 		))
 	}
 

--- a/pkg/component/controller/apiendpointreconciler.go
+++ b/pkg/component/controller/apiendpointreconciler.go
@@ -39,6 +39,10 @@ import (
 // Dummy checks so we catch easily if we miss some interface implementation
 var _ manager.Component = (*APIEndpointReconciler)(nil)
 
+type resolver interface {
+	LookupIPAddr(ctx context.Context, host string) ([]net.IPAddr, error)
+}
+
 // APIEndpointReconciler is the component to reconcile in-cluster API address endpoint based from externalName
 type APIEndpointReconciler struct {
 	logger *logrus.Entry
@@ -47,12 +51,13 @@ type APIEndpointReconciler struct {
 	apiServerPort     int
 	leaderElector     leaderelector.Interface
 	kubeClientFactory kubeutil.ClientFactoryInterface
+	resolver          resolver
 
 	stopCh chan struct{}
 }
 
 // NewEndpointReconciler creates new endpoint reconciler
-func NewEndpointReconciler(nodeConfig *v1beta1.ClusterConfig, leaderElector leaderelector.Interface, kubeClientFactory kubeutil.ClientFactoryInterface) *APIEndpointReconciler {
+func NewEndpointReconciler(nodeConfig *v1beta1.ClusterConfig, leaderElector leaderelector.Interface, kubeClientFactory kubeutil.ClientFactoryInterface, resolver resolver) *APIEndpointReconciler {
 	return &APIEndpointReconciler{
 		logger:            logrus.WithFields(logrus.Fields{"component": "endpointreconciler"}),
 		externalAddress:   nodeConfig.Spec.API.ExternalAddress,
@@ -60,6 +65,7 @@ func NewEndpointReconciler(nodeConfig *v1beta1.ClusterConfig, leaderElector lead
 		leaderElector:     leaderElector,
 		stopCh:            make(chan struct{}),
 		kubeClientFactory: kubeClientFactory,
+		resolver:          resolver,
 	}
 }
 
@@ -105,7 +111,7 @@ func (a *APIEndpointReconciler) reconcileEndpoints(ctx context.Context) error {
 
 	externalAddress := a.externalAddress
 
-	ipAddrs, err := net.DefaultResolver.LookupIPAddr(ctx, externalAddress)
+	ipAddrs, err := a.resolver.LookupIPAddr(ctx, externalAddress)
 	if err != nil {
 		return fmt.Errorf("while resolving external address %q: %w", externalAddress, err)
 	}

--- a/pkg/component/controller/apiendpointreconciler_test.go
+++ b/pkg/component/controller/apiendpointreconciler_test.go
@@ -18,6 +18,7 @@ package controller
 
 import (
 	"context"
+	"net"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -31,25 +32,18 @@ import (
 )
 
 var expectedAddresses = []string{
-	"185.199.108.153",
-	"185.199.109.153",
-	"185.199.110.153",
-	"185.199.111.153",
+	"2001:db8::1",
+	"2001:db8::2",
+	"240.0.0.2",
+	"240.0.0.3",
 }
 
 func TestBasicReconcilerWithNoLeader(t *testing.T) {
 	fakeFactory := testutil.NewFakeClientFactory()
 
-	config := &v1beta1.ClusterConfig{
-		Spec: &v1beta1.ClusterSpec{
-			API: &v1beta1.APISpec{
-				Address:         "1.2.3.4",
-				ExternalAddress: "get.k0s.sh",
-			},
-		},
-	}
+	config := getFakeConfig()
 
-	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: false}, fakeFactory)
+	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: false}, fakeFactory, fakeResolver{})
 
 	ctx := context.TODO()
 	assert.NoError(t, r.Init(ctx))
@@ -66,16 +60,9 @@ func TestBasicReconcilerWithNoLeader(t *testing.T) {
 
 func TestBasicReconcilerWithNoExistingEndpoint(t *testing.T) {
 	fakeFactory := testutil.NewFakeClientFactory()
-	config := &v1beta1.ClusterConfig{
-		Spec: &v1beta1.ClusterSpec{
-			API: &v1beta1.APISpec{
-				Address:         "1.2.3.4",
-				ExternalAddress: "get.k0s.sh",
-			},
-		},
-	}
+	config := getFakeConfig()
 
-	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory)
+	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory, fakeResolver{})
 
 	ctx := context.TODO()
 	assert.NoError(t, r.Init(ctx))
@@ -102,16 +89,9 @@ func TestBasicReconcilerWithEmptyEndpointSubset(t *testing.T) {
 	ctx := context.TODO()
 	_, err = fakeClient.CoreV1().Endpoints("default").Create(ctx, &existingEp, v1.CreateOptions{})
 	assert.NoError(t, err)
-	config := &v1beta1.ClusterConfig{
-		Spec: &v1beta1.ClusterSpec{
-			API: &v1beta1.APISpec{
-				Address:         "1.2.3.4",
-				ExternalAddress: "get.k0s.sh",
-			},
-		},
-	}
+	config := getFakeConfig()
 
-	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory)
+	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory, fakeResolver{})
 
 	assert.NoError(t, r.Init(ctx))
 
@@ -145,15 +125,9 @@ func TestReconcilerWithNoNeedForUpdate(t *testing.T) {
 	_, err := fakeClient.CoreV1().Endpoints("default").Create(ctx, &existingEp, v1.CreateOptions{})
 	assert.NoError(t, err)
 
-	config := &v1beta1.ClusterConfig{
-		Spec: &v1beta1.ClusterSpec{
-			API: &v1beta1.APISpec{
-				Address:         "1.2.3.4",
-				ExternalAddress: "get.k0s.sh",
-			},
-		},
-	}
-	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory)
+	config := getFakeConfig()
+
+	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory, fakeResolver{})
 
 	assert.NoError(t, r.Init(ctx))
 
@@ -188,16 +162,9 @@ func TestReconcilerWithNeedForUpdate(t *testing.T) {
 	_, err := fakeClient.CoreV1().Endpoints("default").Create(ctx, &existingEp, v1.CreateOptions{})
 	assert.NoError(t, err)
 
-	config := &v1beta1.ClusterConfig{
-		Spec: &v1beta1.ClusterSpec{
-			API: &v1beta1.APISpec{
-				Address:         "1.2.3.4",
-				ExternalAddress: "get.k0s.sh",
-			},
-		},
-	}
-	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory)
+	config := getFakeConfig()
 
+	r := NewEndpointReconciler(config, &leaderelector.Dummy{Leader: true}, fakeFactory, fakeResolver{})
 	assert.NoError(t, r.Init(ctx))
 
 	assert.NoError(t, r.reconcileEndpoints(ctx))
@@ -212,4 +179,27 @@ func verifyEndpointAddresses(t *testing.T, expectedAddresses []string, fakeFacto
 	assert.Equal(t, expectedAddresses, endpointAddressesToStrings(ep.Subsets[0].Addresses))
 
 	return ep
+}
+
+type fakeResolver struct{}
+
+func (fr fakeResolver) LookupIPAddr(ctx context.Context, host string) ([]net.IPAddr, error) {
+	return []net.IPAddr{
+		{IP: net.ParseIP("2001:db8::1")},
+		{IP: net.ParseIP("2001:db8::2")},
+		{IP: net.ParseIP("240.0.0.2")},
+		{IP: net.ParseIP("240.0.0.3")},
+	}, nil
+}
+
+func getFakeConfig() *v1beta1.ClusterConfig {
+	return &v1beta1.ClusterConfig{
+		Spec: &v1beta1.ClusterSpec{
+			API: &v1beta1.APISpec{
+				Address:         "240.0.0.1",
+				ExternalAddress: "fake.k0s",
+			},
+		},
+	}
+
 }

